torch/_export/non_strict_utils.py

```diff
@@ -0,0 +1,145 @@
+    fake = mode.from_tensor(t, source=source, symbolic_context=symbolic_context)
+    mode.shape_env.tracked_fakes.append(
+        TrackedFake(fake, source, symbolic_context.constraint_sizes)
+    )
+    return fake
+
+
+def fake_tree(mode, arg, tensor_constraints, source, dim_name_to_sources):
+    if isinstance(arg, list):
+        return [
+            fake_tree(
+                mode,
+                arg,
+                tensor_constraints,
+                GetItemSource(source, i),
+                dim_name_to_sources,
+            )
+            for i, arg in enumerate(arg)
+        ]
+
+    elif isinstance(arg, dict):
+        return {
+            k: fake_tree(
+                mode,
+                arg,
+                tensor_constraints,
+                GetItemSource(source, k),
+                dim_name_to_sources,
+            )
+            for k, arg in arg.items()
+        }
+
+    else:
+        return fakify(mode, arg, tensor_constraints, source, dim_name_to_sources)
+
+
+def make_fake_inputs(nn_module, args, constraints):
+    tensor_constraints: Dict[int, Dict[int, Constraint]] = defaultdict(dict)
+    for constraint in constraints:
+        tensor_constraints[constraint.t_id][constraint.dim] = constraint
+        if constraint.shared is not None:
+            tensor_constraints[constraint.shared.t_id][constraint.shared.dim] = constraint
+
+    code = nn_module.forward.__code__
+    co_fields = {
+        "co_name": code.co_name,
+        "co_filename": code.co_filename,
+        "co_firstlineno": code.co_firstlineno,
+    }
+    with FakeTensorMode(
+        shape_env=ShapeEnv(tracked_fakes=[], co_fields=co_fields)
+    ) as fake_mode:
+        params = inspect.signature(nn_module.forward).parameters
+        dim_name_to_sources: Dict[str, List[Source]] = defaultdict(list)
+        fake_args = tuple(
+            fake_tree(
+                fake_mode, arg, tensor_constraints, LocalSource(x), dim_name_to_sources
+            )
+            for x, arg in zip(params, args)
+        )
+        return fake_mode, fake_args, dim_name_to_sources
+
+
+def make_constraints(fake_mode, dim_name_to_sources, gm):
+    shape_env = fake_mode.shape_env
+    placeholders = [tf.fake for tf in shape_env.tracked_fakes]
+    sources = [tf.source for tf in shape_env.tracked_fakes]
+    constraint_inputs = [tf.constraint_dims for tf in shape_env.tracked_fakes]
+    source_pairs = []
+    for equal_sources in dim_name_to_sources.values():
+        primary_src, *others = equal_sources
+        for src in others:
+            source_pairs.append((primary_src, src))
+    equalities_inputs = EqualityConstraint(
+        source_pairs=source_pairs, warn_only=False
+    )
+    shape_env.produce_guards(
+        placeholders,
+        sources,
+        constraint_inputs=constraint_inputs,
+        equalities_inputs=equalities_inputs,
+        ignore_static=False,
+    )
+    shape_env.frozen = True
+    shape_env.dim_constraints.solve()
```

Comment:
Should we also move the constraint discovery too?

Reply:
Yes I will, separately. It should follow the same pattern as here, but don't want to deal with potential strict test failures in this PR.

Reply:
Wait, I'm not sure what you mean...I was assuming you mean, can we move around steps in strict mode to more closely reflect what's going on here.

As for constraint discovery, well that is essentially `.solve()` + pretty-printing the suggestions. And yes, I forgot to add that. There's actually some non-trivial logic around constraint violation errors that we need to replicate from `eval_frame`, including turning them into user errors, catching them, adding some message related to constraint discovery, etc.

Reply:
Yep I was referring to .solve() + pretty printing the suggestions.